### PR TITLE
feat: engine v1 post process sampled logprobs

### PR DESCRIPTION
**Pull request by RobMcH**
vLLM v1 Engines only support calculating logprobs before any post-processing (temperature, top-k, top-p, etc) is applied. This PR adds the capability to calculate sampled logprobs after the post-processing has run.

Edit: I did an initial test on CPU which seemed to work fine. I'll do more thorough tests tomorrow, opening for review already in the meantime.

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -155,7 +155,8 @@ def __init__(
         self.encoder_cache_size = encoder_cache_size
 
         # Sampler
-        self.sampler = Sampler()
+        self.sampler = Sampler(
```
Comment (robertgshaw2-redhat) [inline vllm/v1/worker/gpu_model_runner.py:158]: I think we also need to apply this to the tpu_model_runner

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 2]
```diff
@@ -82,42 +82,64 @@ def forward_native(
...
             # not needed. This is because `random_sample` does not require
             # CPU-GPU synchronization while `flashinfer_sample` does.
-            return random_sample(probs, generators)
-        return flashinfer_sample(probs, k, p, generators)
+            sample = random_sample(probs, generators)
+            # Logits aren't changed here!
+        else:
+            sample, probs = flashinfer_sample(probs, k, p, generators,
+                                              return_logits)
+            if return_logits:
```
Comment (robertgshaw2-redhat) [inline vllm/v1/sample/ops/topk_topp_sampler.py:118]: why is this needed?

File: vllm/v1/sample/sampler.py [hunk 3]
```diff
@@ -86,49 +92,58 @@ def sample(
...
-    ) -> torch.Tensor:
+        return_logits: bool = False,
+    ) -> tuple[torch.Tensor, Union[torch.Tensor, None]]:
         """Sample logits based on sampling metadata.
 
         The various logits processing functions called in this method
         may update the logits tensor in-place.
         """
-
-        assert not (sampling_metadata.all_greedy
```
Comment (robertgshaw2-redhat) [inline vllm/v1/sample/sampler.py:96]: why does this assertion move?

**Review by robertgshaw2-redhat — CHANGES_REQUESTED** [review]
Thanks for the PR, it seems like a good implementation. Couple thoughts:
- Is there a performance reason why we make the `return_logits` optional? I find that this introduces a lot of if branches in the code that make it hard to read since we continuously have to handle the case of logits=None in the code. Is it possible to just always return the logits and then not have to handle this condition?

- We need to add some tests to this PR for both the flashinfer and non-flashinfer sampler cases

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 4]
```diff
@@ -290,26 +313,32 @@ def flashinfer_sample(
...
+                    probs, uniform_samples, k, deterministic=True))
+        else:
+            # Both top-k and top-p.
+            next_token_ids, success = (
+                flashinfer.sampling.top_k_top_p_sampling_from_probs(
+                    probs, uniform_samples, k, p, deterministic=True))
 
     # NOTE: CPU-GPU synchronization happens here.
-    if not success.all():
+    if return_probs or not success.all():
```
[the remark this answers is not in the document]
Comment (RobMcH) [reply vllm/v1/sample/ops/topk_topp_sampler.py:334]: The branching has an `or` condition, that should implement the behaviour you describe.

File: vllm/v1/sample/sampler.py [hunk 5]
```diff
@@ -46,13 +51,14 @@ def forward(
...
         # Sample the next token.
-        sampled = self.sample(logits, sampling_metadata)
+        sampled, logits = self.sample(logits, sampling_metadata, return_logits)
         # Convert sampled token ids to int64 (long) type to ensure compatibility
         # with subsequent operations that may use these values as indices.
         # This conversion is necessary because FlashInfer sampling operations
         # return int32 (while PyTorch argmax and topk return int64).
         sampled = sampled.long()
-
+        if return_logits:
```
[the remark this answers is not in the document]
Comment (RobMcH) [reply vllm/v1/sample/sampler.py:60]: What is your desired behaviour in this case?

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (RobMcH) [reply vllm/v1/worker/gpu_model_runner.py:158]: Good catch

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 2 above)
Comment (RobMcH) [reply vllm/v1/sample/ops/topk_topp_sampler.py:111]: I just rebased on main, now the function does indeed take the logits instead of probabilities, but the argument still holds for the same reason: we need to modify the logits we return from the sampling to be able to drop any top-k/top-p ignored parts of the vocab.

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RobMcH) [reply vllm/v1/sample/ops/topk_topp_sampler.py]: Just rebased on main and the code path didn't exist anymore, so this is resolved now for sure.

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (RobMcH) [reply vllm/v1/worker/gpu_model_runner.py:135]: I just checked, I think the tpu_model_runner doesn't support logprobs? Or at least it doesn't in the same way the gpu_model_runner does. The sampler **always** returns this
```
        # These are TPU tensors.
        sampler_output = SamplerOutput(
            # The sampled tokens are expanded to 2D tensor with shape
            # [num_requests, 1], where each row represents one generated
            # token per request.
            sampled_token_ids=sampled.unsqueeze(-1),
            logprobs_tensors=None)
        return sampler_output
```

Comment (RobMcH) [reply vllm/v1/worker/gpu_model_runner.py:135]: It's not clear to me why the TPU logprobs are calculated in a different code path (as in, not in the sampler) and I lack context on this. I'd lean to add a TODO for this for whoever is familiar with that code/made that decision, since it would require reworking the logic of the tpu_model_runner (which I assume differs from the gpu_model_runner for a reason).

**Comment by RobMcH** [comment]
@robertgshaw2-redhat I simplified the code now and pushed the return logits check to the outmost layer (the sampler), everything before does return them now. I didn't observe any bad performance implications of this.
